### PR TITLE
mediatek: add LED support for Xiaomi Redmi Router AX6000

### PR DESCRIPTION
**Pull request by HZFrodo**
V1:
1. The LED panel is on SPI1 bus, while SPI1 has a clock issue.
clk_disable_unused() function during the boot process will disable the sel_clk of SPI1 by mistake.
When driver trying to read registers of SPI1, CPU will hang.
So we enable the sel_clk explicitly to make SPI1 full functional.

2. LED panel has 2 physical RGB LEDs, which are controlled by a single wire proprietary protocal.
This device uses MOSI of SPI1 to output the single wire data to both LEDs.
With some kind of reverse-engineering method, I can set RGB colors for each of them finally.
And a new driver, leds-spi-single-wire, was written to light up the LED panel.

V2:
switch to multicolor led driver by @981213

### CODE REVIEW

**Comment by ptpt52** [comment]
mark

**Comment by achterin** [comment]
nice work!

**Comment by lgs2007m** [comment]
Good job! Thanks for you nice work!

**Comment by lalasou** [comment]
very good

**Comment by ziseon** [comment]
Wow, 2 great news (leds support and new partition layouts) for AX6000 at the same time!

**Review by lalasou — APPROVED** [review]

**Review by lalasou — APPROVED** [review]
Very goodI hope it can be quickly merged into the main line.

**Comment by iseeyou** [comment]
mark

**Comment by 981213** [comment]
Is that a WS2812B (NeoPixel) LED? If that's true, it can do way more than just on and off.
Could someone try out the WS2812B protocol on this LED? There are many libraries available for Arduino. (e.g. https://fastled.io/)

**Comment by 981213** [comment]
BTW there's a multicolor LED framework in Linux. If this is an RGB LED I'd prefer it implemented using that instead of registering multiple LEDs with hard-coded color value.

**Comment by dhlun** [comment]
mark

**Comment by HZFrodo** [comment]
> BTW there's a multicolor LED framework in Linux. If this is an RGB LED I'd prefer it implemented using that instead of registering multiple LEDs with hard-coded color value.

Yes, multicolor LED framework is available since kernel version 5.9, but the fact is that we lack of multicolor LED configuration framework in OpenWrt at the momment.
And what is OpenWrt going to do with 24bit colors? If we let our users to select colors at will, then a lot improvements of uci components are required to make multicolor LED work well. Currently, OpenWrt only need four LED states to indicate different status, see: https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh#L6
I am not very familiar to uci releated code, so this should be implemented by some other developer on a suppported target device.

**Comment by 981213** [comment]
We want to avoid drivers which can't be submitted upstream if possible. Registering multiple LEDs that interfere with each other is clearly not acceptable upstream.

> I am not very familiar to uci releated code, so this should be implemented by some other developer on a suppported target device.

Adding a multi_intensity field to led scripts and default config isn't too hard. I could take over this part if you want :)

**Comment by 981213** [comment]
I've written a driver for WS2812B (Tested on Raspberry Pi) and will be sending it upstream. If it's indeed the same protocol, you can also skip the driver part and wait for my patch :)

**Comment by graysky2** [comment]
@HZFrodo - If you try to merge your PR on top of PR#11115 it has conflicts.  Not sure if 11115 will get merged and if so before your PR.  Just FYI.

**Comment by HZFrodo** [comment]
> I've written a driver for WS2812B (Tested on Raspberry Pi) and will be sending it upstream. If it's indeed the same protocol, you can also skip the driver part and wait for my patch :)

That is so nice! Would you create a new PR to make you driver available here? And I can test it on my AX6000. Then I will rebase my PR to utilise your driver

**Comment by HZFrodo** [comment]
> @HZFrodo - If you try to merge your PR on top of PR#11115 it has conflicts. Not sure if 11115 will get merged and if so before your PR. Just FYI.

Since a new driver written by @981213 is coming, a rebase of my PR is required, so PR#11115 could merged first :)

**Comment by 981213** [comment]
@HZFrodo I'll be sending it upstream and get it reviewed there before adding it to OpenWrt.
Meanwhile it would be great if you can test this early version on AX6000 to verify if it's the same protocol. Here's the source code:

[leds-ws2812b-spi.zip](https://github.com/openwrt/openwrt/files/10063628/leds-ws2812b-spi.zip)

and the current dt binding looks like this:

```dts
#include <dt-bindings/leds/common.h>
		spi@0 {
			status = "okay";
                        #address-cells = <1>;
                        #size-cells = <0>;
			ws2812: ws2812b@0 {
				#address-cells = <1>;
				#size-cells = <0>;
				compatible = "worldsemi,ws2812b-spi";
				reg = <0>;
				spi-max-frequency = <2850000>;
                                led@0 {
                                        reg = <0>;
                                        label = "rgb:led0";
					color-index = <LED_COLOR_ID_RED LED_COLOR_ID_GREEN LED_COLOR_ID_BLUE>;
					default-intensity = <10 50 90>;
                                        color = <LED_COLOR_ID_RGB>;
                                        function = LED_FUNCTION_STATUS;
					function-enumerator = <0>;
                                };
				led@1 {
                                        reg = <1>;
                                        color = <LED_COLOR_ID_RGB>;
                                        function = LED_FUNCTION_STATUS;
					function-enumerator = <1>;
                                };
			};
		};
```

**Comment by HZFrodo** [comment]
> [leds-ws2812b-spi.zip](https://github.com/openwrt/openwrt/files/10063628/leds-ws2812b-spi.zip)
@981213 Your new driver works fine on AX6000, I can set different colors and different triggers.
Once this driver is added to OpenWrt, I will rebase my PR to utilise your driver, thanks.

**Review by lgs2007m — APPROVED** [review]

**Comment by graysky2** [comment]
@HZFrodo @981213 - How can one try the driver?

**Comment by graysky2** [comment]
@HZFrodo - I would like to try this PR but have already merged [PR#11115](https://github.com/openwrt/openwrt/pull/11115).  Is it safe to just change `/mt7986a-xiaomi-redmi-router-ax6000.dts/mt7986a-xiaomi-redmi-router-ax6000.dtsi/` in this PR as-is?  I believe that is the target for the patch once you merge PR#11115.  Thanks.

**Comment by ptpt52** [comment]
@graysky2 
you may pick from https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c
and https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27

**Comment by graysky2** [comment]
@ptpt52 - thanks, works great!

**Review by missing233 — APPROVED** [review]

**Review by sunnyday666 — COMMENTED** [review]
This is perfect.

**Review by sunnyday666 — APPROVED** [review]

**Review by lalasou — COMMENTED** [review]
ok

**Review by lalasou — COMMENTED** [review]
ok

**Review by iseeyou — COMMENTED** [review]
ok

**Comment by 981213** [comment]
Please stop spamming this PR with empty review. We don't merge PR directly via Github. The "At least 6 approving reviewers" is only there to prevent accidental merge button clicking.

**Comment by 981213** [comment]
Upstreaming the driver takes a bit longer than I thought...
@HZFrodo Would you mind taking my v4 patch here and update this PR?
https://patches.linaro.org/project/linux-leds/patch/20221212045558.69602-4-gch981213@gmail.com/

I'll then merge this PR and take care of replacing it with the final version when it got accepted upstream.
DT maintainer upstream suggests that default-intensity should not be a driver-specific property so I dropped it. The LED can only be white for now. I'll do something about the color part in the userspace after merging this PR.

**Comment by graysky2** [comment]
@981213 - Have you seen these from @ptpt52 :

> you may pick from https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c
and https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27

I picked those two commits and have blue, green, red, white, and yellow.

85 ti

**Comment by graysky2** [comment]
OK... so in the future, are you recommending that I not use the two commits from you but instead use this?

**Comment by 981213** [comment]
> @981213 - Have you seen these from @ptpt52 :
> 
> > you may pick from [x-wrt@e614fad](https://github.com/x-wrt/x-wrt/commit/e614fad37ed72f5d152c31212060dad131055d1c)
> > and [x-wrt@3c1b8dc](https://github.com/x-wrt/x-wrt/commit/3c1b8dc8ca72d8c1d008b455f420fb8b68844a27)
> 
> I picked those two commits and have blue, green, red, white, and yellow.

It hard-coded some color values for this RGB LED chip and this approach will definitely be rejected on Linux upstream. I'd like to avoid carrying a local patch when there is an upstream-ready approach available (implementing its driver with the multicolor led interface).

**Comment by HZFrodo** [comment]
> Upstreaming the driver takes a bit longer than I thought... @HZFrodo Would you mind taking my v4 patch here and update this PR? https://patches.linaro.org/project/linux-leds/patch/20221212045558.69602-4-gch981213@gmail.com/
> 
> I'll then merge this PR and take care of replacing it with the final version when it got accepted upstream. DT maintainer upstream suggests that default-intensity should not be a driver-specific property so I dropped it. The LED can only be white for now. I'll do something about the color part in the userspace after merging this PR.

No problem, I will take your pending patch and rework this PR :-)

**Review by lalasou — APPROVED** [review]

**Comment by HZFrodo** [comment]
@981213 PR reworked.
Tested on my AX6000, all good.

**Comment by graysky2** [comment]
@HZFrodo - do I just need to pull this PR on top of HEAD to verify functionality?

**Comment by HZFrodo** [comment]
> @HZFrodo - do I just need to pull this PR on top of HEAD to verify functionality?

Yes. But if you are using the previous version of my PR and do a sysupgrade with configuration retained, you have to edit the LED config section in /etc/config/system to use the new LED label, or LED will not light up.

File: target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts
Comment (981213) [inline target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts:220]: Does it really need a cs-gpios property?

Comment (HZFrodo) [reply target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dts:220]: Of course not, copied it from spi0 by mistake, I will delete it, thanks

**Comment by 981213** [comment]
I turned the LED patch into a standalone package and split the commits. I've also dropped the heartbeat trigger because I'd prefer the status indicator to behave like other OpenWrt devices by default.
Merged. Thanks!

**Comment by 981213** [comment]
BTW if anyone want to set a color for their router now, you can write the RGB value you want to the multi_intensity sysfs interface in /etc/rc.local

**Comment by graysky2** [comment]
> I've also dropped the heartbeat trigger because I'd prefer the status indicator to behave like other OpenWrt devices by default.

Oh no, I love the heartbeat settings.  Any chance to re-add?  Users are of course free to switch them on or not.

**Comment by 981213** [comment]
> Oh no, I love the heartbeat settings. Any chance to re-add? Users are of course free to switch them on or not.

You can easily add that trigger in luci if you want it.

**Pull request closed without merging**
